src/v/utils/abort_source_with_timeout.cc

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "utils/abort_source_with_timeout.h"
+
+#include "vlog.h"
+
+#include <seastar/core/timed_out_error.hh>
+#include <seastar/util/log.hh>
+
+#include <numeric>
+
+abort_source_with_timeout::abort_source_with_timeout(
+  ss::abort_source& as, ss::lowres_clock::duration timeout)
+  : _abort_on_expiry{ss::lowres_clock::now() + timeout}
+  , _sources({as, _abort_on_expiry.abort_source()}) {
+    subscribe_to_sources();
+}
+
+ss::abort_source& abort_source_with_timeout::abort_source() {
+    return _derived_as;
+}
+
+void abort_source_with_timeout::subscribe_to_sources() {
+    _subscriptions[0] = _sources[0].get().subscribe(
+      [this](const std::optional<std::exception_ptr>& opt_eptr) noexcept {
+          if (!_derived_as.abort_requested()) {
+              _derived_as.request_abort_ex(opt_eptr.value_or(
+                std::make_exception_ptr(ss::abort_requested_exception{})));
+          }
+      });
```

Comment:
just an idea, haven't tested it. 
maybe if this triggers, there is a window of opportunity for the abort_on_expiry to trigger before being destroyed, and this is causing the timed_out_error in the new test case.

to test this, I think you can wrap abort_on_expiry in an optional, to reset in this handler. it should work and if it does not trigger ci it could be the solution, or you can manually reimplement the logic by using a seastar::timer instead of abort_on_expiry

Reply:
I just realized that no abort source is passed into that test, so we are using `p.get_shared_future(deadline)` which does throw `timed_out_error` on timeout. I have updated the test now.